### DISCUSSION TITLE
Session refresh with extra options

### DISCUSSION THREAD

**Discussion by Impro02** [sqlalchemy/sqlalchemy#12053] [Usage Questions] [answered]
Hi all,

Given the following models User and Address:

```
class Base(DeclarativeBase):
    pass

class Address(Base):
    __tablename__ = "ADDRESS"

    id: Mapped[int] = mapped_column(
        "ID",
        Integer,
        primary_key=True,
        index=True,
    )
    street: Mapped[str] = mapped_column(
        "STREET",
        String,
        index=True,
    )
    zip_code: Mapped[str] = mapped_column(
        "ZIP_CODE",
        Integer,
        index=True,
    )
    user_id: Mapped[int] = mapped_column(
        "USER_ID",
        Integer,
        ForeignKey("USER.ID"),
    )

    user: Mapped["User"] = relationship(
        "User",
        back_populates="addresses",
        lazy="joined",
    )


class User(Base):
    __tablename__ = "USER"

    id: Mapped[int] = mapped_column(
        "ID",
        Integer,
        primary_key=True,
        index=True,
    )
    email: Mapped[str] = mapped_column(
        "EMAIL",
        String,
        unique=True,
        index=True,
    )

    addresses: Mapped[List["Address"]] = relationship(
        "Address",
        back_populates="user",
        lazy="joined",
        cascade="all, delete-orphan",
    ) 
``` 

Now, we want to load the user data for a given user_id, but only load the linked addresses with a criteria, here the zip codes:

``` 
zip_codes = ["00741", "02132"]

stmt = select(User).options(
    lazyload(User.adresses.and_(Adress.zip_code.in_(zip_codes )))
).filter(User.id== 1)

user= session.execute(stmt).unique().scalar_one_or_none()
```

Later, if we make some update on the user, and apply the session refresh to retrieve the up-to-date user, all the addresses are loaded without taking into account our zip codes criteria. Is there a trick to keep the extra criteria while refreshing the session?

```
session.refresh(user)
```

**Comment by zzzeek** [comment, marked as answer]
a refresh implies an expire so that will un-do those loader options.

a more open ended way to refresh is to use [session.get()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.get) with populate existing - but you have to have it load those attributes in order to overwrite with the new value:

    session.get(user.id, options=[immediateload(User.adresses.and_(Adress.zip_code.in_(zip_codes ))], populate_existing=True)
